contracts/adapters/UniswapAdapter.sol

```diff
@@ -0,0 +1,233 @@
+        uint deadline = block.timestamp + deadlineBuffer;
+        IERC20 tokenIn = IERC20(addr1);
+        address[] memory path = new address[](2);
+        path[0] = addr1;
+        path[1] = addr2;
+        if (tokenIn.allowance(address(this), address(router)) == 0){
+                SafeERC20.safeApprove(tokenIn, address(router), type(uint256).max);
+        }
+
+        router.swapExactTokensForTokens(amount,amountOutMin, path, address(this), deadline);
+
+    }
+
+    function _addLiquidity(address token1,address token2,uint256 amount1,uint256 amount2, IUniswapV2Router02 router) internal {
+        uint deadline = block.timestamp + deadlineBuffer;
+        IERC20 toAdd = IERC20(token1);
+        IERC20 toAdd2 = IERC20(token2);
+        if (toAdd.allowance(address(this), address(router)) < amount1){
+                SafeERC20.safeApprove(toAdd, address(router), type(uint256).max);
+        }
+        if (toAdd2.allowance(address(this), address(router)) < amount2){
+                SafeERC20.safeApprove(toAdd2, address(router), type(uint256).max);
+        }
+
+        router.addLiquidity(token1,token2, amount1, amount2, 0, 0, address(this), deadline);
+    }
+
+    //By the time this function is called the user bal should be in wbtc
+    //calculates optimal swap amt for minimal leftover funds and buys Digg
+    // Provides liquidity and transfers lp token to msg.sender
+    function _buyLp( uint256 userWbtcBal, Exchange exchange, address traderAccount, uint256 minDiggAmtOut)  internal{
+
+        if (exchange == Exchange.Uniswap){
+                (uint112 reserve_amt,,) = IUniswapV2Pair(wbtcDiggUniswap).getReserves();
+                uint256 trade_amt = getSwapAmt(reserve_amt,userWbtcBal);
+                uint256 startingDiggBal = IERC20(diggAddress).balanceOf(address(this));
+                convertTokenToToken(wbtcAddress, diggAddress, trade_amt, minDiggAmtOut, exchange);
+                uint256 startLpBal = wbtcDiggUniswap.balanceOf(address(this));
+                addLiquidity(wbtcAddress, diggAddress, userWbtcBal, IERC20(diggAddress).balanceOf(address(this)).sub(startingDiggBal), exchange);
+                wbtcDiggUniswap.transfer(traderAccount, wbtcDiggUniswap.balanceOf(address(this)).sub(startLpBal));
+        }
+        else if (exchange == Exchange.Uniswap){
+                (uint112 reserve_amt,,) = IUniswapV2Pair(wbtcDiggSushiswap).getReserves();
+                uint256 trade_amt = getSwapAmt(reserve_amt,userWbtcBal);
+                uint256 startingDiggBal = IERC20(diggAddress).balanceOf(address(this));
+                convertTokenToToken(wbtcAddress, diggAddress, trade_amt, minDiggAmtOut, exchange);
+                uint256 startLpBal = wbtcDiggSushiswap.balanceOf(address(this));
+                addLiquidity(wbtcAddress, diggAddress, userWbtcBal, IERC20(diggAddress).balanceOf(address(this)).sub(startingDiggBal), exchange);

```

Comment:
Using this [idea here](https://github.com/ribbon-finance/structured-products/pull/87/files#r592986267), we can avoid querying `IERC20(diggAddress).balanceOf(address(this))`.

```solidity
uint diggAmount = convertTokenToToken(wbtcAddress, diggAddress, trade_amt, minDiggAmtOut, exchange);
addLiquidity(wbtcAddress, diggAddress, userWbtcBal, diggAmount, exchange);
```